trinity/extensibility/plugin_manager.py

```diff
@@ -0,0 +1,79 @@
+from argparse import (
+    ArgumentParser
+)
+import logging
+from typing import (
+    Iterable,
+    List,
+    Union,
+)
+
+from trinity.extensibility.events import (
+    BaseEvent,
+    PluginStartedEvent,
+)
+from trinity.extensibility.plugin import (
+    BasePlugin,
+)
+
+
+class PluginManager:
+    """
+    The plugin manager is responsible to register, keep and manage the life cycle of any available
+    plugins.
+
+      .. note::
+
+        This API is very much in flux and is expected to change heavily.
+    """
+
+    def __init__(self) -> None:
+        self._plugin_store: List[BasePlugin] = []
+        self._started_plugins: List[BasePlugin] = []
+        self._logger = logging.getLogger("trinity.extensibility.plugin_manager.PluginManager")
+
+    def register(self, plugins: Union[BasePlugin, Iterable[BasePlugin]]) -> None:
+        """
+        Register one or multiple instances of :class:`~trinity.extensibility.plugin.BasePlugin`
+        with the plugin manager.
+        """
+
+        new_plugins = [plugins] if isinstance(plugins, BasePlugin) else plugins
+
+        for val in new_plugins:
+            self._plugin_store.append(val)
```

Comment:
loop can be replaced with `self._plugin_store.extend(new_plugins)`

This will go away completely if you choose to go with only supporting a single plugin at a time for registration.

Reply:
Ah, nice. Let's use `extend` then and continue to support batch registration if you don't mind :unicorn: